Environment group merging and previews

Xetera opened the pull request:
## Pull request type

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

- [ ] Bugfix
- [x] Feature
- [ ] Other (please describe):

## Pull request checklist

- [ ] If it's a backend change, tests for the changes have been added and `go test ./...` runs successfully from the root folder.
- [x] If it's a frontend change, Prettier has been run
- [ ] Docs have been reviewed and added / updated if needed

## What is the current behavior?

When trying to load environment variables, all existing variables are overwritten by the variables in the selected environment group.

![current implementation](https://user-images.githubusercontent.com/24978328/124369286-d867dd00-dc72-11eb-91c1-9f2dbd1b2ca6.png)

Closes #821

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue.

Issue Number: N/A

-->

## What is the new behavior?

- When selecting environment variables, you can view the contents of the variables
- You can see any conflicting variables that are both defined in the current context and incoming group 

![image](https://user-images.githubusercontent.com/24978328/124369346-a99e3680-dc73-11eb-92ef-fed8ec1a1c78.png)

![image](https://user-images.githubusercontent.com/24978328/124369566-955b3900-dc75-11eb-8dff-e99e5c2c14ed.png)


<!-- Please describe the behavior or changes that are being added by this PR. -->

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

## Technical Spec/Implementation Notes

A few missing types were added.

`SaveButton` implementation was tweaked to prevent the longer status texts from overflowing outside the modal padding by adding a `left` position and `justify-content: space-between`.

The modal main content was split in half to allow adding dynamic content without having to vertically resize the modal.

abelanger5 reviewed (commented):
Hey @Xetera, thanks so much for this PR! Just tested locally and the core functionality is working great. I requested @jnfrati to take a closer look at the code. 

A few high-level considerations, interested to hear your thoughts on these:
- One edge case here is the "secret" environment variables. These get loaded into the frontend as "dummy" environment variables with the data `PORTERSECRET_<name-of-env-group>`, so the "Replaced by..." text is slightly confusing. Perhaps we should add a case to make it clear that this is a "secret" env variable?
- Perhaps increase the size of the modal slightly? The split modal is super useful with the env group previews for what is getting replaced, but it's starting to feel a bit cramped.

jnfrati reviewed (commented):
It looks good to me! Although I would expect @jusrhee to have comments on how it looks like, but we can achieve those on another PR if needed.

Xetera commented:
I think the env group should be taking up 100% of the width before one is selected so we don't have to spend 50% of the space on a placeholder text.

Also yes now that the modal has more responsibility it should definitely be taking up a little more space. Should I just increase it in to take up more space in pixels or is there a bigger refactoring that should happen here?

abelanger5 commented:
> I think the env group should be taking up 100% of the width before one is selected so we don't have to spend 50% of the space on a placeholder text.

Yes, I agree. 

> Also yes now that the modal has more responsibility it should definitely be taking up a little more space. Should I just increase it in to take up more space in pixels or is there a bigger refactoring that should happen here?

Yep, just increasing the pixel value is fine! I'm guessing the modals will be some of the last items that are refactored, so don't worry about it.

Xetera commented:
Hopefully this addresses the issues. The preview still looks a bit out of place and kind of ugly, especially when you have real environment variables that end up wrapping but I didn't want to add a syntax highlighting lib without consulting the maintainers.

The `<hr />` between the preview and warnings still feels a bit out of place but I couldn't figure out how to make that work.

![image](https://user-images.githubusercontent.com/24978328/125155007-b89a5280-e165-11eb-9c10-9f730ced81e4.png)

![image](https://user-images.githubusercontent.com/24978328/125155944-87bd1c00-e16b-11eb-946f-8439079bef20.png)

abelanger5 commented on dashboard/src/components/SaveButton.tsx:
```diff
@@ -74,17 +74,22 @@ const LoadingGif = styled.img`
 `;
 
+const StatusContainer = styled.div`
+  flex: 1;
```
Setting the `StatusContainer` to `flex` causes the status updates on `ExpandedChart` to render completely on the left of the chart:

![image](https://user-images.githubusercontent.com/25448214/125176160-58c09d80-e19f-11eb-8623-10b86a56037b.png)

Is `flex` necessary here/perhaps we parameterize where we want the status text to show?

abelanger5 commented on dashboard/src/components/SaveButton.tsx:
```diff
@@ -74,17 +74,22 @@ const LoadingGif = styled.img`
 
   max-width: 500px;
-  white-space: nowrap;
+  white-space: break-spaces;
```
This has some unintended consequences I think, this is the reason `white-space` is set to `nowrap `: 

![image](https://user-images.githubusercontent.com/25448214/125175997-0a5ecf00-e19e-11eb-96fe-76f89e2f63ad.png)

(our errors aren't normally this long, but there is a way to trigger it by setting the raw Helm values incorrectly on purpose. In this case, I manually set `cpu: 100mASDF` in the helm values tab.)

I'm sure there are other solutions, but the ones that come to mind are: set the maximum height to 2 lines, trim the text that gets rendered for the status, or change `white-space` back to `nowrap` and shorten the error (since the modal is larger now, the longest error is pretty close to 1 line).

abelanger5 reviewed (changes requested):
Awesome, it's working great, really cleans up the environment var flow @Xetera 💯

Sorry that I didn't catch these last time, but tweaking the `SaveButton` has some consequences in other parts of the platform. See the review comments.

abelanger5 commented on dashboard/src/main/home/modals/LoadEnvGroupModal.tsx:
```diff
@@ -96,7 +102,55 @@ export default class LoadEnvGroupModal extends Component<PropsType, StateType> {
+            <ClashingKeyDefinitions>
+              <ClashingKeyLabel>Old</ClashingKeyLabel>
+              <ClashingKeyValue>
+                {this.props.existingValues[key] || emptyValue}
```
Shouldn't this use `formattedEnvironmentValue` as well? When testing, the "Old" keys are still showing as `PORTER_SECRET`: 

![image](https://user-images.githubusercontent.com/25448214/125176445-6840e600-e1a1-11eb-99d5-68089d6a23e5.png)

Xetera replied:
Totally, I have no idea how that slipped by :grimacing:

Xetera commented:
I moved the save button changes to a separate PR (#878) because it was starting to clutter this one. Everything else related to environment merging is ready on my side.

abelanger5 commented:
Perfect, tested one last time and merged! Should be on the hosted version within the hour. I'll take a look at the `SaveButton` PR. 

Thanks @Xetera!
